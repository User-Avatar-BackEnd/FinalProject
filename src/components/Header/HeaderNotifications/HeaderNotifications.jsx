import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBell } from '@fortawesome/free-solid-svg-icons';
import API from '../../../config/API';
import useComponentVisible from '../../../hooks/useComponentVisible';

import styles from './HeaderNotifications.module.scss';

const HeaderNotifications = () => {
  const [notifications, setNotifications] = useState([])
  const { ref, isComponentVisible, setIsComponentVisible } = useComponentVisible(false);

  const token = localStorage.getItem('AUTH_TOKEN')

  useEffect( () => {
    if (isComponentVisible) {
      API({
        method: 'get',
        url: '/account/invites',
        headers: {
          Authorization: `Bearer ${token}`
        }
      })
        .then(response => {
          setNotifications(response.data)
        })
        .catch(() => {
          setNotifications([])
        })
    } else {
      setNotifications([])
    }
  }, [isComponentVisible])

  const toggleList = () => {
    setIsComponentVisible(!isComponentVisible)
  }

  const acceptInvite = (e) => {
    inviteResponse(1, e.target.getAttribute('data-id'))
  }

  const declineInvite = (e) => {
    inviteResponse(-1, e.target.getAttribute('data-id'))
  }

  const inviteResponse = (status, id) => {
    API({
      method: 'patch',
      url: `/account/invites/${id}/?status=${status}`,
      headers: {
        Authorization: `Bearer ${token}`
      }
    })
      .then(() => {
        setNotifications(notifications.filter(item => {
          return item.id !== Number(id)
        }))
      })
      .catch(error => {
        console.log(error)
      })
  }

  return (
    <div className={styles.HeaderNotifications} ref={ref}>
      <div className={styles.notificationsIcon} onClick={toggleList}>
        <FontAwesomeIcon icon={faBell}/>
      </div>
      {isComponentVisible &&
        <div className={styles.notificationsList}>
          {notifications.length === 0
            ? <h3>You have no notifications yet...</h3>
            : ''
          }
          {notifications.slice(0, 3).map(item => {
            return <div className={styles.notification} key={item.id}>
              <span>{item.inviter?.login} invited you to the board {item.board.title}</span>
              <div className={styles.controlButtons}>
                <div className={styles.accept} data-id={item.id} onClick={acceptInvite}>
                  Accept
                </div>
                <div className={styles.decline} data-id={item.id} onClick={declineInvite}>
                  Decline
                </div>
              </div>
            </div>
          })}
          {notifications.length > 3
            ? <div className={styles.seeAllLink}>
              <Link
                to={{
                  pathname: '/profile',
                  state: {
                    page: 2
                  }
                }}
                onClick={toggleList}
              >
                See all invites
              </Link>
            </div>
            : ''
          }

        </div>
      }
    </div>
  )
}

export default HeaderNotifications;