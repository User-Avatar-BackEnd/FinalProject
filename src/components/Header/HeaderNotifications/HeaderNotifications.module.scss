.HeaderNotifications {
  position: relative;

  & .notificationsIcon {
    font-size: 26px;
    line-height: 50px;
    cursor: pointer;
    transition: color .3s;

    &:hover {
      color: orange;
    }
  }

  & .notificationsList {
    position: absolute;
    width: 300px;
    min-height: 220px;
    background-color: #eaeaea;
    left: -270px;
    box-shadow: rgba(0, 0, 0, 0.35) 0 5px 15px;
    border-radius: 10px;
    z-index: 2;

    &:before {
      content: "";
      border: solid transparent;
      position: absolute;
      right: 10px;
      bottom: 100%;
      border-bottom-color: #eaeaea;
      border-width: 10px;
      margin-left: 0;
    }

    & h3 {
      color: #000;
    }

    & .notification {
      display: flex;
      flex-direction: column;
      justify-content: space-between;
      width: 90%;
      height: 50px;
      background-color: #f3f3f3;
      margin: 10px auto;
      border-radius: 10px;
      font-size: 12px;
      padding: 3px;
      box-sizing: border-box;
      color: #000;
      box-shadow: rgba(100, 100, 111, 0.2) 0 7px 29px 0;

      & .controlButtons {
        display: flex;
        justify-content: space-between;
        font-size: 14px;
        width: 60%;
        margin: 0 auto;

        & .accept, & .decline {
          height: 20px;
          width: 60px;
          text-align: center;
          line-height: 20px;
          border-radius: 3px;
          color: #fff;
          cursor: pointer;
          transition: background-color .3s;
        }

        & .accept {
          background-color: seagreen;

          &:hover {
            background-color: darken(green, 5%);
          }
        }

        & .decline {
          background-color: indianred;

          &:hover {
            background-color: darken(red, 10%);
          }
        }
      }
    }

    & .seeAllLink {
      text-align: center;

      & a {
        color: orange;
        transition: color .3s;

        &:hover {
          color: blue;
        }
      }
    }
  }
}
