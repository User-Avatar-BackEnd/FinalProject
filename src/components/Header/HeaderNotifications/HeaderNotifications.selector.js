const selector = state => ({
  notifications: state.user.notifications,
  loading: state.user.isNotificationsLoading
})

export default selector;