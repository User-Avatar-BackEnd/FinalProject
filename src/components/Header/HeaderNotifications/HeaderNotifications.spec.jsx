import React from 'react';
import { shallow } from 'enzyme';

import HeaderNotifications from './HeaderNotifications';

describe('HeaderNotifications component', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<HeaderNotifications />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should match snapshot', () => {

    expect(wrapper).toMatchSnapshot();
  });
});