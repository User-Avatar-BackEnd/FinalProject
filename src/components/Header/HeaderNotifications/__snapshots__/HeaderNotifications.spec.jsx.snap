// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderNotifications component should match snapshot 1`] = `
<div
  className="HeaderNotifications"
>
  <div
    className="notificationsIcon"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f0f3",
            "M224 512c35.32 0 63.97-28.65 63.97-64H160.03c0 35.35 28.65 64 63.97 64zm215.39-149.71c-19.32-20.76-55.47-51.99-55.47-154.29 0-77.7-54.48-139.9-127.94-155.16V32c0-17.67-14.32-32-31.98-32s-31.98 14.33-31.98 32v20.84C118.56 68.1 64.08 130.3 64.08 208c0 102.3-36.15 133.53-55.47 154.29-6 6.45-8.66 14.16-8.61 21.71.11 16.4 12.98 32 32.1 32h383.8c19.12 0 32-15.6 32.1-32 .05-7.55-2.61-15.27-8.61-21.71z",
          ],
          "iconName": "bell",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
</div>
`;
